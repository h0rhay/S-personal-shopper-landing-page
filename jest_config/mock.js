export const mockTheme = {
  themeType: 'dark',
  breakpoints: {
    small: '(max-width: 812px)',
    medium: '(min-width: 813px)',
    large: '(min-width: 960px)',
    deviceSwitchSm: '(max-width: 896px)',
    deviceSwitchLg: '(min-width: 897px)',
  },
  typography: {
    mobile: {
      headerOne: {
        fontSize: '3rem',
      },
      headerTwo: {
        fontSize: '1.75rem',
      },
      displayHeading: {
        fontSize: 'calc(1.5rem * 2)',
      },
      itemHeading: {
        fontSize: '1.125rem',
      },
    },
    desktop: {
      headerOne: {
        fontSize: '4rem',
      },
      headerTwo: {
        fontSize: '2rem',
      },
      displayHeading: {
        fontSize: 'calc(4rem * 2)',
      },
      itemHeading: {
        fontSize: '1.125rem',
      },
    },
  },
  text: '#ffffff',
  primary: '#212121',
  primaryAccent: '#2E2E2E',
  accent: '#ffe256',
  shadow: 'rgba(0, 0, 0, 0.4)',
};
