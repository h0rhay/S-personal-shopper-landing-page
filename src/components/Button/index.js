import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'gatsby';
import styled from 'styled-components';

const isAnchorOrLink = (href, to, onClick, target, rel) => {
  if (href) {
    return {
      href,
      target,
      rel,
    };
  }
  if (onClick) {
    return {
      onClick,
    };
  }
  return {
    to,
  };
};

const ButtonStyles = styled.div`
  a,
  button {
    display: inline-block;
    border: none;
    outline: 1px solid ${({ theme }) => theme.accent};
    padding: 1rem 5.5vmin;
    margin: 0;
    text-decoration: none;
    text-transform: none;
    background: ${({ theme }) => theme.primary};
    color: ${({ theme }) => theme.text};
    line-height: 1;
    cursor: pointer;
    text-align: center;
    transition: background 250ms ease-in-out, transform 150ms ease;
    white-space: nowrap;
    -webkit-appearance: none;
    -moz-appearance: none;

    &:hover,
    &:focus {
      color: ${({ theme }) => theme.primary};
      background: ${({ theme }) => theme.accent};
      outline: 1px solid ${({ theme }) => theme.primary};
    }

    &:focus {
      outline: 1px solid ${({ theme }) => theme.text};
      outline-offset: -4px;
    }

    &:active {
      transform: scale(0.99);
    }
  }
`;

const Button = ({ to, href, onClick, children, target, rel, type, shopperId }) => {
  const ActionComponent = to ? Link : href ? 'a' : 'button';
  const url = href ? new URL(href) : null;
  if (shopperId) {
    url.searchParams.append('utm_source', 'clientelling');
    url.searchParams.append('utm_medium', 'referral');
    url.searchParams.append('utm_campaign', `${shopperId}_clientellingmicrosite_na_na_na_na`);
    url.searchParams.append('utm_term', 'na');
    url.searchParams.append('utm_content', 'na');
    url.searchParams.append('POR', 'Y');
  }
  const actions = isAnchorOrLink(url, to, onClick, target, rel);
  const componentType = ActionComponent === 'button' ? type || 'button' : undefined;
  return (
    <ButtonStyles>
      <ActionComponent {...actions} onClick={onClick} type={componentType}>
        {children}
      </ActionComponent>
    </ButtonStyles>
  );
};

Button.propTypes = {
  children: PropTypes.string,
  href: PropTypes.string,
  onClick: PropTypes.func,
  rel: PropTypes.string,
  shopperId: PropTypes.number,
  target: PropTypes.string,
  to: PropTypes.string,
  type: PropTypes.string,
};

export default Button;
