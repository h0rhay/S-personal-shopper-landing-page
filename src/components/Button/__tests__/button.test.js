import React from 'react';
import { render, screen } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { mockTheme } from '../../../../jest_config/mock';
import Button from '../index';

jest.mock('gatsby-link', () => <a href="mock-test">Mock Link</a>);

const renderComponent = ({ theme, props }) =>
  render(
    <ThemeProvider theme={theme}>
      <Button {...props} />
    </ThemeProvider>
  );

describe('Button component with "to" provided', () => {
  const props = {
    children: 'Shop now',
    to: '/shop',
  };

  it('should match screenshot', () => {
    const { container } = renderComponent({ theme: mockTheme, props });
    expect(container).toMatchSnapshot();
  });
});

describe('Button component with "href" provided', () => {
  const props = {
    children: 'Shop now',
    href: 'https://www.selfridges.com/GB/en/cat/zimmermann-concert-wool-midi-dress_R03806044/?previewAttribute=CREAM',
    shopperId: 325565,
    target: '_blank',
    rel: 'noopener noreferrer',
    to: null,
  };

  it('should match screenshot', () => {
    const { container } = renderComponent({ theme: mockTheme, props });
    expect(container).toMatchSnapshot();
  });

  it('the url should contain tracking params', () => {
    renderComponent({ theme: mockTheme, props });
    const button = screen.getByText(/shop now/i);
    expect(button.href).toContain('utm_source=clientelling');
  });
});

describe('Button component with "onClick" provided', () => {
  const props = {
    children: 'Shop now',
    onClick: () => {},
  };

  it('should match screenshot', () => {
    const { container } = renderComponent({ theme: mockTheme, props });
    expect(container).toMatchSnapshot();
  });
});
