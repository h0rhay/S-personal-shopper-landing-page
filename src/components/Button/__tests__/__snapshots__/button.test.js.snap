// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component with "href" provided should match screenshot 1`] = `
<div>
  <div
    class="sc-bdvvtL gGhPvv"
  >
    <a
      href="https://www.selfridges.com/GB/en/cat/zimmermann-concert-wool-midi-dress_R03806044/?previewAttribute=CREAM&utm_source=clientelling&utm_medium=referral&utm_campaign=325565_clientellingmicrosite_na_na_na_na&utm_term=na&utm_content=na&POR=Y"
      rel="noopener noreferrer"
      target="_blank"
    >
      Shop now
    </a>
  </div>
</div>
`;

exports[`Button component with "onClick" provided should match screenshot 1`] = `
<div>
  <div
    class="sc-bdvvtL gGhPvv"
  >
    <button
      type="button"
    >
      Shop now
    </button>
  </div>
</div>
`;

exports[`Button component with "to" provided should match screenshot 1`] = `
<div>
  <div
    class="sc-bdvvtL gGhPvv"
  >
    <a
      href="/shop"
    >
      Shop now
    </a>
  </div>
</div>
`;
